### ISSUE TITLE
Improve configuration of dev container .gitconfig

### ISSUE THREAD

**Issue by chrmarti** [microsoft/vscode-remote-release#4632]
Thank you for responding @chrmarti, I appreciate you discussing this even if you didn't necessarily agree with it.

I really appreciate the tip regarding `devcontainer.json`. That is non-obvious but it works great!

I still question why the local configuration is imported by default, but with a working _shippable_ setting to disable that behaviour, I'm good.

I'm now running the following shell script as my `postCreateCommand`:

```sh
git config --global core.editor "code --wait"
git config --global diff.tool vscode
git config --global difftool.vscode.cmd "code --wait --diff \$LOCAL \$REMOTE"
git config --global merge.tool vscode
git config --global mergetool.vscode.cmd "code --wait \$MERGED"
yarn install
```

Maybe in the future this extension would consider setting more than just `credential.helper` by default; "text editor sets itself as text editor" seems like fair game in terms of behaviour, and would probably improve ease of use for most folks. It would also overwrite any incompatible imported settings.

_Originally posted by @aaronadamsCA in https://github.com/microsoft/vscode-remote-release/issues/4603#issuecomment-792959197_

**Comment by posita** [comment]
I may have misunderstood, but +100 to not overwriting the container user's `.gitignore` as a default. As far as I can tell, this was the same beef as with #4603. Silently overwriting a bunch of stuff on behalf of the user by default is not only disrespectful, but it's hard to debug. Please either turn it off by default or make it stupid simple to detect and disable without requiring protracted investigations by end users.

**Comment by chrmarti** [comment]
`.gitconfig` is only copied into the container when there is no `.gitconfig` in the container yet. The `credential.helper` is then overwritten with a helper that forwards the credentials from the host. Both can be controlled through user settings.

**Comment by posita** [comment]
Which user settings? Link?

**Comment by chrmarti** [comment]
<img width="902" alt="image" src="https://user-images.githubusercontent.com/9205389/155158900-a15322fd-4ab0-4d45-9239-9c5f43b3405e.png">

**Comment by chrmarti** [comment]
Check the Remote tab. That will store the setting in your container's home folder. (Or disconnect from the remote and the setting will be under the User tab.)

**Comment by posita** [comment]
Nothing under the remote tab either.

<img width="359" alt="Screen Shot 2022-02-23 at 08 05 26" src="https://user-images.githubusercontent.com/222581/155335054-a2b5cc76-29af-47d7-8797-1688214732fe.png"> <img width="330" alt="Screen Shot 2022-02-23 at 08 05 32" src="https://user-images.githubusercontent.com/222581/155335059-f9882d1f-61bc-4049-9f90-45361661f43f.png">

**Comment by aidan-melen** [comment]
> <img alt="image" width="902" src="https://user-images.githubusercontent.com/9205389/155158900-a15322fd-4ab0-4d45-9239-9c5f43b3405e.png">

thank you kind sir

**Comment by posita** [comment]
Still not seeing this. @aidan-melen, how did you get there?

**Comment by aaronadamsCA** [comment]
@posita If searching settings for "git config" gives you no results, something's up with your environment well beyond the topic of this issue. A clean install of VS Code should return at least 4 results for that search term just from built-in extensions.

**Comment by posita** [comment]
Okay, I reinstalled and now have those settings. `Git Credential Helper Config Location` _**only**_ appears in `Remote -Containers`, though, not `User`. When I set it to `global` and restart my container, my container user's `~/.gitconfig` still gets modified, though. Not sure if that's a separate bug, but it's certainly _**not**_ honoring the settting.

**Comment by aaronadamsCA** [comment]
I remember finding that `copyGitConfig` only worked for me when I put it into my `devcontainer.json` file, like this:

```json
{
  "settings": {
    "remote.containers.copyGitConfig": false
  }
}
```

It's possible `gitCredentialHelperConfigLocation` has the same limitation, I'm not sure.

**Comment by posita** [comment]
I tried that. It's not honored. This does not appear to have any effect:

``` python
{
  "settings": {
    "remote.containers.gitCredentialHelperConfigLocation": "global"
  }
}
```

I have verified that the setting _does_ propagate to `~/.vscode-server/data/Machine/settings.json` on the container side.

**Comment by chrmarti** [comment]
Note that `"remote.containers.gitCredentialHelperConfigLocation": "global"` is the default and copies the `.gitconfig` to the user's home folder. `global` and `system` are Git's terminology for the gitconfig's location.

**Comment by aaronadamsCA** [comment]
Note also that this issue can probably be closed, since it was asking for VS Code to make itself the Git text editor, and that seems to happen now, at least for me.

**Comment by moander** [comment]
I see #6124 (git email/name is not copied) if I run `git config --global ...` in the `postCreateCommand` script like @chrmarti do.

**Comment by chrmarti** [comment]
The `.gitconfig` is not copied if one already exists in the container.

The copying is done after `postCreateCommand` and any dotfiles install script run, so if `postCreateCommand` or the dotfiles installs script create a `.gitconfig` in the container (e.g., as part of `git config --global ...`), the local `.gitconfig` will not be copied.

**Comment by teohhanhui** [comment]
There must be a way to *avoid* modifying the user's `.gitconfig` file, right? :thinking:

**Comment by chrmarti** [comment]
@teohhanhui See this user setting:
![Image](https://github.com/microsoft/vscode-remote-release/assets/9205389/6073afd0-0458-4740-9cda-5c3622d032d1)

**Comment by teohhanhui** [comment]
@chrmarti That's on by default, and it's talking about *copying* the host's `.gitconfig` into the container, no? How does that relate to *modifying* the host's `.gitconfig`?

**Comment by chrmarti** [comment]
@teohhanhui Which modification are you referring to?

**Comment by teohhanhui** [comment]
@chrmarti Correct me if I'm wrong, but from what I've observed the extension adds its own credential helper to `~/.gitconfig`, which causes lots of problems, especially when it fails to clean up after itself.

Also, it does `chown` as well...

It'd really be preferable for the extension to only modify the project's `.git/config` in the case of credential helpers.

**Comment by chrmarti** [comment]
@teohhanhui It sounds like you have your local home folder mounted in the container. You could change the following user setting to `system` (there is no `project` at the moment):
![Image](https://github.com/microsoft/vscode-remote-release/assets/9205389/a27ccbef-2cbb-4361-9621-f7274a05e3d3)

**Comment by teohhanhui** [comment]
> It sounds like you have your local home folder mounted in the container. 

Yes, it's a very common setup. I'm using `distrobox`, but `toolbox` does the same as well.

**Comment by branic** [comment]
https://github.com/microsoft/vscode-remote-release/issues/4632#issuecomment-1495924470
> The `.gitconfig` is not copied if one already exists in the container.
> 
> The copying is done after `postCreateCommand` and any dotfiles install script run, so if `postCreateCommand` or the dotfiles installs script create a `.gitconfig` in the container (e.g., as part of `git config --global ...`), the local `.gitconfig` will not be copied.

Is this documented anywhere? This is the only place I've been able to find this information and think it would be very valuable to have in the documentation as it would have saved me a bunch of time trying to figure out why my .gitconfig wasn't being copied into the container.
